### PR TITLE
tracking and strip DQM : trigger selection via DB

### PR DESCRIPTION
**Pull request by mtosi**
this PR is meant to --finally-- enable the trigger selection via DB

I run runTheMatrix -l limited, and it runs smoothly

in order to make use of the DB,
I use the following setting in the DQM config

import CondCore.DBCommon.CondDBSetup_cfi
process.dbInput = cms.ESSource( "PoolDBESSource"
, CondCore.DBCommon.CondDBSetup_cfi.CondDBSetup
, connect  = cms.string( 'sqlite_file:/afs/cern.ch/work/t/tosi/public/DQM/checkDB/CMSSW_8_0_4/src/DQM/TrackerCommon/test/AlCaRecoTriggerBits_TrackerDQM.db' )
, toGet    = cms.VPSet(
    cms.PSet(
      record = cms.string( 'AlCaRecoTriggerBitsRcd' )
    , tag    = cms.string( 'AlCaRecoTriggerBits_TrackerDQM_v1' )
    , label  = cms.untracked.string( 'SiStripDQMTrigger' )
    )
  )
)
process.es_prefer = cms.ESPrefer("PoolDBESSource","dbInput")

where the sqlite file is like
# 
# AlCaRecoTriggerBits settings for IOV 132000-132999:
# 

trigger list key: 'SiStripDQM_L1'
paths:
'NOT L1Tech_BSC_halo_beam2_inner.v0', 'NOT L1Tech_BSC_halo_beam2_outer.v0', 'NOT L1Tech_BSC_halo_beam1_inner.v0', 'NOT L1Tech_BSC_halo_beam1_outer.v0',
'NOT (L1Tech_BSC_splash_beam1.v0 AND NOT L1Tech_BSC_splash_beam2.v0)', 'NOT (L1Tech_BSC_splash_beam2.v0 AND NOT L1Tech_BSC_splash_beam1.v0)'
# 

trigger list key: 'SiStrip_HLT'
paths:
'HLT_ZeroBias_v_', 'HLT_HIZeroBias_v_'
# 

trigger list key: 'SiStrip_L1'
paths:
'L1Tech_BPTX_plus_AND_minus.v0', 'L1_ZeroBias'
# 

trigger list key: 'Tracking_HLT'
paths:
'HLT_ZeroBias_v*'
# 

we will need to update the payload, indeed

@mmusich @fioriNTU @rovere @VinInn

### CODE REVIEW

**Comment by mmusich** [comment]
please test

**Comment by mtosi** [comment]
could @deguio and/or @vanbesien please report the matrix which is run @tier0 ?
it would be useful to cross check which sequence (and module) is run and when ...
thanks

**Comment by mmusich** [comment]
@miatosi, @deguio @vanbesien as far I can understand this if deployed,  calls for an update of the Global Tag, namely to append a new IOV that introduces the new keys `SiStrip_HLT`, `SiStrip_L1` and `Tracking_HLT` in the tag for record `AlCaRecoTriggerBitsRcd` with label `SiStripDQMTrigger`.
Questions are:
1) is this run anywhere else than at Tier0 ?
2) related to 1) do we have a way of testing the new paradigma short of Tier0 replay?

**Comment by mtosi** [comment]
looking at the log, it is not clear how this PR is related to these errors :(

18 13 12 11 3 0 0 tests passed, 0 4 0 0 0 0 0 failed
140.53, 1000.0, 1001.0, 1003.0: exit: 0 16640 0 0 0

Begin processing the 1st record. Run 191226, Event 556818698, LumiSection 398 at 20-Apr-2016 12:41:28.466 CEST
----- Begin Fatal Exception 20-Apr-2016 12:41:42 CEST-----------------------
An exception of category 'NoRecord' occurred while
   [0] Processing run: 191226 lumi: 398 event: 556818698
   [1] Running path 'reconstruction_step'
   [2] Calling event method for module L1FastjetCorrectorProducer/'ak4PFL1FastjetCorrector'
Exception Message:
No "JetCorrectionsRecord" record found in the EventSetup for synchronization value
Run: 191226 LuminosityBlock: 398 Event: 0 Time: 5731090697768189300
 Please add an ESSource or ESProducer that delivers such a record.
----- End Fatal Exception -------------------------------------------------
%MSG-w BeamFitter:  AlcaBeamMonitor:AlcaBeamMonitor@endLumi  20-Apr-2016 12:41:42 CEST Run: 191226 Lumi: 398
No event read! No Fitting!

**Comment by mmusich** [comment]
@mtosi it's not. It's the IB that has issues... (my fault incidentally)

**Comment by mmusich** [comment]
please test

**Comment by deguio** [comment]
ciao,
I think that  WF 1000.0 and 1001.0 are a good approximation of what runs at the T0. they are run in the jenkins tests and complete successfully.

@mtosi I would like to know more about what the intentions are.

**Comment by mmusich** [comment]
ciao,
then I am puzzled. 

We don't have the new keys that Mia introduced 

https://github.com/cms-sw/cmssw/pull/14161/files#diff-5834b6b487acac21668a25b15da4154dR9
https://github.com/cms-sw/cmssw/pull/14161/files#diff-5834b6b487acac21668a25b15da4154dR19

 in the IOV probed by 1000.0 and 1001.0.
How comes it passes the tests?

**Comment by mtosi** [comment]
I've just re-run 1000.0 and 1001.0 tests
both tests run on
/store/data/Run2011A/MinimumBias/RAW/v1/000/165/121/0699429A-B37F-E011-A57A-0019B9F72D71.root

1000.0 runs
process.dqmoffline_step = cms.Path(process.DQMOfflineCommon)
so, only the L1 selection used by the SiStrip module should be called

while 1001.0 runs
process.dqmoffline_step = cms.Path(process.DQMOfflineCommonSiStripZeroBias)
so, both L1 selection used by SiStrip module and HLT selection used by SiStrip and Tracking modules are used

running in both 1000.0 and 1001.0 I find
l1LogicalExpressionsCache_:
and in 1001.0 there are also
hltLogicalExpressionsCache_:

this because we are missing the label in the DB (as we know)
but, as reported in the GenericTriggerEventFlag twiki [https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideGenericTriggerEventFlag]
"
 cms.string l1DBKey (optional):
    Tag of the record n the database, where IOV-based L1 algorithms are found. This record overwrites the configuration parameter l1Algorithms.
    ==l1Algorithms== remains valid if:
        l1DBKey is not present (database access switched off);
        *database access fails. *
"
and
"
 cms.string hltDBKey (optional):
Tag of the record in the database, where IOV-based HLT paths are found. This record overwrites the configuration parameter hltPaths.
==hltPaths== remains valid if:

    hltDBKey is not present (database access switched off);
    *database access fails*
"
right ?

**Comment by mmusich** [comment]
please test

File: CommonTools/TriggerUtils/src/GenericTriggerEventFlag.cc [hunk 1]
```diff
@@ -84,6 +84,7 @@ GenericTriggerEventFlag::GenericTriggerEventFlag( const edm::ParameterSet & conf
         gtEvmInputToken_ = iC.mayConsume< L1GlobalTriggerEvmReadoutRecord >( gtEvmInputTag_ );
       }
       if ( config.exists( "gtDBKey" ) ) gtDBKey_ = config.getParameter< std::string >( "gtDBKey" );
+      std::cout << "[GenericTriggerEventFlag::GenericTriggerEventFlag] gtDBKey: " << gtDBKey_ << std::endl;
```
Comment (mmusich) [inline CommonTools/TriggerUtils/src/GenericTriggerEventFlag.cc]: cout is not thread safe. any reason why not to use LogInfo?

**Comment by mtosi** [comment]
I've just realized I already committed the GenericTriggerEventFlag I'm working on (sorry)
I'll be back shortly w/ the final file

**Comment by slava77** [comment]
-1
reco mods are only in CommonTools/TriggerUtils/src/GenericTriggerEventFlag.cc 
and it looks like a wrong version is committed (with debugs, before cleanup)

**Comment by slava77** [comment]
@cmsbuild please test

File: CommonTools/TriggerUtils/src/GenericTriggerEventFlag.cc [hunk 2]
```diff
@@ -137,10 +137,16 @@ void GenericTriggerEventFlag::initRun( const edm::Run & run, const edm::EventSet
     if ( onL1_ && l1DBKey_.size() > 0 ) {
       const std::vector< std::string > exprs( expressionsFromDB( l1DBKey_, setup ) );
       if ( exprs.empty() || exprs.at( 0 ) != configError_ ) l1LogicalExpressionsCache_ = exprs;
+      for ( unsigned iExpr = 0; iExpr < l1LogicalExpressionsCache_.size(); ++iExpr ) {
+	std::string l1LogicalExpression( l1LogicalExpressionsCache_.at( iExpr ) );
```
Comment (slava77) [inline CommonTools/TriggerUtils/src/GenericTriggerEventFlag.cc]: I'm missing the point of these lines: it's a loop that assigns a value to a temporary that's never used.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test
the last one failed due to 
``` socket.timeout: timed out ```

**Comment by slava77** [comment]
@mtosi  
was the payload updated (mentioned in the end of the PR description) .. or what else is missing ?

I see that there are new warning messages like in run2 2015 workflow 134.911
```
%MSG-w GenericTriggerEventFlag:   SiStripMonitorCluster:SiStripMonitorClusterBPTX@streamBeginRun  04-May-2016 09:51:47 CEST Run: 256677
No logical expressions found under key SiStrip_L1 in 'AlCaRecoTriggerBitsRcd'
%MSG
%MSG-w GenericTriggerEventFlag:   MonitorTrackResiduals:MonitorTrackResiduals@streamBeginRun  04-May-2016 09:51:48 CEST Run: 256677
No logical expressions found under key SiStrip_HLT in 'AlCaRecoTriggerBitsRcd'
%MSG
```

The SiStrip/MechanicalView plots are dead
e.g. in 134.911
wf134911_sistrip_diagpos
the same behavior is seen in run1 and run2 data and MC processing.

**Comment by mtosi** [comment]
the payload has not been updated yet, I would say
I let @musich comments on this
the warnings are there because of this
and we need them on in case of mismatch
(once the payload is fine, they should gone)
the same applies to the DQM plot
now, it relies only on DB trigger config, therefore in case of missing/mismatching the plot is empty, but the warning is triggered
in the DQM the warnings are monitored globally (plots should be under the LogMessage dir -by heart-) the idea is to make use of the historic DQM developed by the tracker DQM for catching such kind of issues

**Comment by mmusich** [comment]
@slava77 and @mtosi,
indeed the Global Tags were not updated yet. 
I asked Mia explicitly to configure things like this (also adding the warnings) exactly in order to be able to spot where we need to update. Before the last consolidation of the configuration enhancing the verbosity level it was easy not to spot inconsistencies. 
I'll arrange with Mia on which tags need to be udpated with wich configuration. 
For starters: do we want the same trigger selection applied on MC as well?

**Comment by mtosi** [comment]
ciao
I might be wrong, but I remember that in both relval sample and MC workflows
the DQM matrix makes use of the tracker/tracking DQM sequence w/ trigger selection
=> it should not be needed

the only situation when it will be needed on MC (but only NuGun and DY in leptons, I would say) is the data/MC exercise .. which is a special scheme, though

  mia

**Comment by slava77** [comment]
I am interested in these DQM plots in relvals for comparisons enough to not yet sign this PR (despite rather minimal changes in reco code).

**Comment by mtosi** [comment]
these ones https://goo.gl/Dq9Pjy ?

**Comment by slava77** [comment]
I meant the effect of this PR on the relval matrix https://cloud.githubusercontent.com/assets/4676718/15014113/531c74fa-1204-11e6-8094-35267b7f0907.png
Not something that's already in present relvals.

**Comment by slava77** [comment]
@mtosi please clarify on the status of this PR

**Comment by slava77** [comment]
-1
merge conflicts need to be resolved

**Comment by mtosi** [comment]
ciao
I'll update this PR asap
but, please let me understand what needs to be clarified, it is not clear to me

in addition,
the GenericTriggerEventFlag needs to be updated in order to handle the new L1 offline code
I'm working on it, but I have a doubt
do I have to make it back compatible ?
I mean, make it able to run on stage2, stage1 and legacy L1 ?
thanks much

**Comment by slava77** [comment]
On 5/20/16 5:30 AM, mia tosi wrote:
> ciao
> I'll update this PR asap
> but, please let me understand what needs to be clarified, it is not
> clear to me
> 
> in addition,
> the GenericTriggerEventFlag needs to be updated in order to handle the
> new L1 offline code
> I'm working on it, but I have a doubt
> do I have to make it back compatible ?
> I mean, make it able to run on stage2, stage1 and legacy L1 ?
> thanks much


My 2c on this:
We occasionally may need to reprocess old data.
DQM may be needed  to validate the reprocessing quality.
So, it's better to be able to support the past logic in some way.


		--slava

**Comment by mtosi** [comment]
if I'm not wrong
when we do rereco we almost always fail in taking the proper DQM sequence @deguio 
=> I agree w/ you @slava77 , but I would ask to @deguio to "guarantee" that we will have the proper sequence run on the corresponding PD ;)

while, could you please tell me what I should clarify ?
thanks

**Comment by mtosi** [comment]
in addition
@slava77 , the work I did and tested so far is
https://github.com/cms-sw/cmssw/compare/CMSSW_8_0_X...mtosi:from-CMSSW_8_0_7_patch2_fixGenericTriggerEventFlag?expand=1
as you see, I dropped the previous L1GtUtils and make use of the new L1GlobalUtils
+
I modify the l1LogicalExpressions_ vector
(I explicitly add the erase)
in order to avoid an "always true" in case the trigger provided by the configuration does not match any one in the L1 menu
I did it because w/ the current setting of the SiStripMonitorCluster there is indeed such an issue

said that,
I would say that what I did should do the update job
but, for making it "back-compatible", 
I would exploit the new input parameters I added
in order to distinguish between the 2 cases ... 
or are there any other solutions ?

anyhow, we should modify the sequences by using the era, right ?
could you point me to the proper code ?

thanks

**Comment by slava77** [comment]
@mtosi 
Thank you very much for trying more with this on the side CommonTools/TriggerUtils
It's outside of the tracking code.

@mulhearn @rekovic 
I wish we had a common interface supported by L1 software to get access to bits, algo names, and prescales for the totality of CMS data taking.
Are the consumers of L1 information forced to figure it out on their own 
or is there something recommended to be able to handle access to L1 trigger info transparently?
Please advise.
Maybe this can continue in the hypernews.

**Comment by mtosi** [comment]
ciao

sorry, but I'm really confused
in order to understand which is the issue on the merging and then fix it
I've just done the following

cd CMSSW_8_1_X_2016-05-24-2300/src/
cmsenv
git cms-merge-topic 14161

and I got
From .
 * [new ref]                    -> merge-attempt
From https://github.com/cms-sw/cmssw
 * [new ref]         refs/pull/14161/head -> cms-sw/refs/pull/14161/head
Switched to branch 'merge-attempt'
Already up-to-date.
M	DQM/TrackingMonitor/interface/LogMessageMonitor.h
M	DQM/TrackingMonitor/src/LogMessageMonitor.cc
M	DQM/TrackingMonitorSource/python/LogMessageMonitor_cff.py
M	DQM/TrackingMonitorSource/python/TrackingSourceConfig_Tier0_cff.py
Switched to branch 'from-CMSSW_8_0_4_triggerGenericEventFilterViaDB'
Checking out these packages: 0

=> I do not see any issue w/ the merging of the PR in 81x

what am I missing ?
thanks much

**Comment by mtosi** [comment]
then, if you agree,
I will add the updates I developed in this branch
https://github.com/cms-sw/cmssw/compare/CMSSW_8_1_X...mtosi:from-CMSSW_8_0_7_patch2_fixGenericTriggerEventFlag

hope it is fine
thanks

**Comment by makortel** [comment]
@mtosi If I do
```
cmsrel CMSSW_8_1_X_2016-05-24-2300
cd CMSSW_8_1_X_2016-05-24-2300/src
cmsenv
git cms-merge-topic 14161
```
I get
```
Auto-merging DQM/TrackingMonitor/src/TrackingMonitor.cc
Auto-merging DQM/TrackingMonitor/python/TrackingMonitor_cfi.py
Auto-merging DQM/TrackerMonitorTrack/src/MonitorTrackResiduals.cc
Auto-merging DQM/SiStripMonitorTrack/src/SiStripMonitorTrack.cc
Auto-merging DQM/SiStripMonitorClient/python/SiStripSourceConfigTier0_cff.py
CONFLICT (content): Merge conflict in DQM/SiStripMonitorClient/python/SiStripSourceConfigTier0_cff.py
Automatic merge failed; fix conflicts and then commit the result.
```
presumably because of #14569.

**Comment by mtosi** [comment]
thanks Matti
I started from scratch, and I copied and pasted your commnads
I'm still not getting the error, but

From .
 * [new ref]                    -> merge-attempt
Switched to branch 'merge-attempt'
Already up-to-date.
M	DQM/TrackingMonitor/interface/LogMessageMonitor.h
M	DQM/TrackingMonitor/src/LogMessageMonitor.cc
M	DQM/TrackingMonitorSource/python/LogMessageMonitor_cff.py
M	DQM/TrackingMonitorSource/python/TrackingSourceConfig_Tier0_cff.py
Switched to branch 'from-CMSSW_8_0_4_triggerGenericEventFilterViaDB'
Checking out these packages: 0

puzzled ...

**Comment by mmusich** [comment]
@mtosi still conficts. Have you rebased?

**Comment by slava77** [comment]
On 5/26/16 2:46 AM, mia tosi wrote:
> then, if you agree,
> I will add the updates I developed in this branch
> CMSSW_8_1_X...mtosi:from-CMSSW_8_0_7_patch2_fixGenericTriggerEventFlag
> <https://github.com/cms-sw/cmssw/compare/CMSSW_8_1_X...mtosi:from-CMSSW_8_0_7_patch2_fixGenericTriggerEventFlag>
> 
> hope it is fine

It looks like it covers both uGT and GT.
Good.
Please, go ahead.

**Comment by slava77** [comment]
On 5/26/16 3:19 AM, mia tosi wrote:
> thanks Matti
> I started from scratch, and I copied and pasted your commnads
> I'm still not getting the error, but
> 
> From .
> 
>   * [new ref] -> merge-attempt Switched to branch 'merge-attempt'
>     Already up-to-date. M
>     DQM/TrackingMonitor/interface/LogMessageMonitor.h M
>     DQM/TrackingMonitor/src/LogMessageMonitor.cc M
>     DQM/TrackingMonitorSource/python/LogMessageMonitor_cff.py M
>     DQM/TrackingMonitorSource/python/TrackingSourceConfig_Tier0_cff.py
>     Switched to branch 'from-CMSSW_8_0_4_triggerGenericEventFilterViaDB'
>     Checking out these packages: 0
> 

Switched to branch 'from-CMSSW_8_0_4_triggerGenericEventFilterViaDB'
this doesn't look right.
git cms-merge-topic 14161
should put you back on from-CMSSW_8_1_X_2016-05-24-2300
are you sure that cmsenv worked? check $CMSSW_BASE
(git cms-merge-topic  makes all changes in $CMSSW_BASE/src,
regardless of which directory you are currently in)

**Comment by mtosi** [comment]
the rebase is taking ~forever :(
=> I've just prepared a new branch (starting from scratch)
I'm going to close this PR and open a new one ... cross the fingers
then, I will have to do a new PR for backporting all these updates in 80x as well ..... double cross the fingers ;)

**Pull request closed without merging**